dicom_validator/validator/iod_validator.py

```diff
@@ -84,54 +102,55 @@ def _validate_module(self, module, module_name):
                 tag_id = self._tag_id(tag_id_string)
                 if tag_id in self._dataset:
                     message = self._incorrect_tag_message(tag_id,
-                                                          'not allowed', None)
+                                                          'not allowed')
                     errors.setdefault(message, []).append(tag_id_string)
         else:
-            for tag_id_string, attribute in module_info.items():
-                result = self._validate_attribute(self._tag_id(tag_id_string),
-                                                  attribute)
-                if result is not None:
-                    errors.setdefault(result, []).append(tag_id_string)
+            errors.update(self._validate_attributes(module_info, False))
         return errors
 
-    def _log_module_required(self, module_name, required, allowed,
-                             condition_dict):
-        msg = 'Module "' + module_name + '" is '
-        msg += ('required' if required
-                else 'optional' if allowed else 'not allowed')
-        if condition_dict:
-            condition = Condition.read_condition(condition_dict)
-            if condition.type != 'U':
-                msg += ' due to condition:\n  '
-                msg += condition.to_string(self._dict_info)
-        self.logger.debug(msg)
+    def _validate_attributes(self, attributes, report_unexpected_tags):
+        errors = {}
 
-    def _incorrect_tag_message(self, tag_id, error_kind, condition_dict):
-        msg = 'Tag {} is {}'.format(tag_name_from_id(tag_id, self._dict_info),
-                                    error_kind)
-        if condition_dict:
-            condition = Condition.read_condition(condition_dict)
-            if condition.type != 'U':
-                msg += ' due to condition:\n  '
-                msg += condition.to_string(self._dict_info)
-        return msg
+        for tag_id_string, attribute in attributes.items():
+            tag_id = self._tag_id(tag_id_string)
+
+            result = self._validate_attribute(tag_id, attribute)
+            if result is not None:
+                errors.setdefault(result, []).append(tag_id_string)
+            
+            self._dataset_stack[-1].unexpected_tags.discard(tag_id)
+
+            if "items" in attribute:
+                data_elem = self._dataset_stack[-1].dataset.get_item(tag_id)
+                if data_elem is None:
+                    continue
+                if data_elem.VR != "SQ":
+                    raise RuntimeError(f"Not a sequence: {data_elem}")
+                for i, sq_item_dataset in enumerate(data_elem.value):
+                    self._dataset_stack.append(
+                        DatasetStackItem(sq_item_dataset, tag_id_string))
+                    errors.update(self._validate_attributes(attribute["items"], True))
```

Comment:
I think I didn't validate sequence items, because I thought we cannot use the same rules for items as for the (root) dataset, and I'm still not completely sure if this is correct... can you explain your reasoning here?

Reply:
Why cannot we use the same/similar rules? It is true that right now the validator's ConditionParser is limited for certain cases (e.g. "This attrib is required in the Nth sequence item only if the 1st item had it"). But that does not stop the validator to "check what it can". That was my idea at least.

Reply:
Actually this looks fine - sorry, I had misread the code the first time and also forgot that there was the "items" entry that allows exactly that. As I wrote, it has been some time... Would be nice to a have a test for that, but I guess that can be done later after some test refactoring.